pandas/io/html.py

```diff
@@ -258,161 +260,263 @@ def _parse_td(self, obj):
+        footer_rows = self._parse_tfoot_tr(table_html)
+
+        def row_is_all_th(row):
+            return all(self._equals_tag(t, 'th') for t in
+                       self._parse_td(row))
+
+        if not header_rows:
+            # The table has no <thead>. Move the top all-<th> rows from the
+            # <tbody> to the <thead>. (This is a common case because many
+            # tables in the wild have no <thead> or <tfoot>
+            while body_rows and row_is_all_th(body_rows[0]):
+                header_rows.append(body_rows.pop(0))
+
+        if not footer_rows:
+            # The table has no <tfoot>. Treat last all-<th> rows as footers.
+            while body_rows and row_is_all_th(body_rows[-1]):
+                # .insert(), not .append(): we're moving "bottom of <tbody>" to
+                # "top of <tfoot>"
+                footer_rows.insert(0, body_rows.pop())
+
+        header = self._expand_colspan_rowspan(header_rows)
+        body = self._expand_colspan_rowspan(body_rows)
+        footer = self._expand_colspan_rowspan(footer_rows)
+
         return header, body, footer
 
-    def _parse_raw_thead(self, table):
-        thead = self._parse_thead(table)
-        res = []
-        if thead:
-            trs = self._parse_tr(thead[0])
-            for tr in trs:
-                cols = lmap(self._text_getter, self._parse_td(tr))
-                if any(col != '' for col in cols):
-                    res.append(cols)
-        return res
-
-    def _parse_raw_tfoot(self, table):
-        tfoot = self._parse_tfoot(table)
-        res = []
-        if tfoot:
-            res = lmap(self._text_getter, self._parse_td(tfoot[0]))
-        return np.atleast_1d(
-            np.array(res).squeeze()) if res and len(res) == 1 else res
-
-    def _parse_raw_tbody(self, table):
-        tbodies = self._parse_tbody(table)
-
-        raw_data = []
-
-        if tbodies:
-            for tbody in tbodies:
-                raw_data.extend(self._parse_tr(tbody))
-        else:
-            raw_data.extend(self._parse_tr(table))
+    def _expand_colspan_rowspan(self, rows):
+        """
+        Given a list of <tr>s, return a list of text rows that copy cell
+        text across rowspans/colspans.
+
+        Parameters
+        ----------
+        rows : list of <tr>s
+
+        Returns
+        -------
+        list of list
+            Each returned row is a list of str text.
+        """
 
-        return self._parse_raw_data(raw_data)
+        all_texts = []  # list of rows, each a list of str
+        remainder = []  # list of (index, text, nrows)
+
+        for tr in rows:
+            texts = []  # the output for this row
+            next_remainder = []
+
+            index = 0
+            tds = self._parse_td(tr)
+            for td in tds:
+                # Append texts from previous rows with rowspan>1 that come
+                # before this <td>
+                while remainder and remainder[0][0] <= index:
```

Comment:
Thanks for the comment before about iteration. Can you help point me to where `remainder` first gets values populated for it? I'm still somewhat concerned about the number of mutations done to that variable. 

I have the feeling some of this can be better and more safely expressed using enumerate but direction on where to begin with this would be helpful

Reply:
`remainder` is initialized on line 447 and flipped with `next_remainder` on line 485.

It's valid to worry about an evil web page forcing lots of `remainder` mutations ... to a point. If I were generating a website to foil Pandas, here's what I'd do:

```html
<table>
  <tr>
    <td colspan="2147483647">This will crash your puny parser.</td>
  </tr>
</table>
```

I don't expect this to spin forever because of mutations: I expect it to crash with out-of-memory as it grows an enormous list.

Real-world tables will be fine. Usually colspan/rowspan are only in header cells, and there are rarely many items here.

Big-O analysis: for a table with `N` output cells (N = WxH), each `.pop(0)` generates a cell. That means there are at most `N` list-removals, meaning at most `N` list-insertions. Each insertion is `O(1)` but each removal is `O(W)`. (According to https://wiki.python.org/moin/TimeComplexity, list pop is O(k).) That removal is the slowest operation, so running time is `O(NW)`.

Shall I switch to a `collection.deque` to make it `O(N)`?

Reply:
Incidentally: I, too, started by writing `enumerate()`. But that fell apart. The index of a cell in the output depends on two inputs: both `td`'s index in `row` and the number of items in `remainder` that pre-empt it.

An alternate strategy is be to make `remainder` a list of mostly-`None`, sometimes-tuple values and modify them in-place. That, too, would be `O(N)` ... but it wouldn't involve as many allocations.

Reply:
Have to think through that in more detail. I'm less concerned about the choice of container; I just feel like there's a more readable and concise way to approach this that would be easier for future contributors / maintainers to use